File: Proxy02.js
//Objeto Real - Classe Produto
class Produto{
  constructor(nome, preco){
    this.nome = nome;
    this.preco = preco;
  }

  exibirDetalhes(){
    console.log(`Produto: ${this.nome}, Preco: ${this.preco.toFixed(2)}`);
  }
}
//FAZER OS 3 EM PY

//Proxy - Classe Produto
class ProxyProduto{
  constructor(produto){
    this.produto = produto;
  }

  exibirDetalhes(){
    console.log(`Aguardando autenticacao...`);
    this.autenticar();
    this.produto.exibirDetalhes();
  }


  autenticar(){
    console.log(`Autenticando usuario....`);
    //Logica de autenticacao aqui.....
    console.log(`Ususario autenticado com sucesso...`);
  }
}



//Cliente - Uso do Proxy
const produto1 = new Produto("Camiseta", 39.90);
const proxyProduto = new ProxyProduto(produto1);
proxyProduto.exibirDetalhes();

File: FAÇADE (1)/proxy02.py
class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco

    def exibir(self):
        print(f"Produto: {self.nome}, Preço: {self.preco:.2f}")

class ProxyProduto:
    def __init__(self, produto):
        self.produto = produto

    def exibirDetalhes(self):
        print("Aguardando Autenticação...")
        self.autenticar()
        self.produto.exibir()

    def autenticar(self):
        print("Autenticando Usuário...")
        print("Usuário Autenticado com Sucesso...")

produto1 = Produto("Camiseta", 39.90)
proxyproduto = ProxyProduto(produto1)
proxyproduto.exibirDetalhes()
